src/media/media.ts

```diff
@@ -0,0 +1,1041 @@
+        kind,
+        requestedDeviceId: deviceId,
+        sdkConfigDefaultDeviceId,
+        sessionInfos
+      });
+    }
+
+    return foundDevice ? foundDevice.deviceId : undefined;
+  }
+
+  /**
+   * Get the current _cached_ media devices
+   */
+  getDevices (): MediaDeviceInfo[] {
+    return this.getState().devices;
+  }
+
+  /**
+   * Get a copy of the current media state
+   */
+  getState (): SdkMediaState {
+    return cloneDeep(this.state);
+  }
+
+  /**
+   * Get the current _cached_ audio devices
+   */
+  getAudioDevices (): MediaDeviceInfo[] {
+    return this.getState().audioDevices;
+  }
+
+  /**
+   * Get the current _cached_ video devices
+   */
+  getVideoDevices (): MediaDeviceInfo[] {
+    return this.getState().videoDevices;
+  }
+
+  /**
+   * Get the current _cached_ output devices
+   */
+  getOutputDevices (): MediaDeviceInfo[] {
+    return this.getState().outputDevices;
+  }
+
+  /**
+   * This will return all active media tracks that
+   *  were created by the SDK
+   */
+  getAllActiveMediaTracks (): MediaStreamTrack[] {
+    return Array.from(this.allMediaTracksCreated.values());
+  }
+
+  /**
+   * Look through the cached devices and match based on
+   *  the passed in track's `kind` and `label`. Returns the
+   *  found device or `undefined` if the device could not be found.
+   * @param track media stream track with the label to search for
+   */
+  findCachedDeviceByTrackLabel (track?: MediaStreamTrack): MediaDeviceInfo | undefined {
+    if (!track) return;
+    return this.getDevices().find(d =>
+      d.label === track.label && `${d.kind.substr(0, 5)}` === track.kind
+    );
+  }
+
+  /**
+   * Look through the cached devices and match based on
+   *  the passed in track's `kind` and `label`. Returns the
+   *  found device or `undefined` if the device could not be found.
+   * @param track media stream track with the label to search for
+   */
+  doesDeviceExistInCache (device?: MediaDeviceInfo): boolean {
+    if (!device) return false;
+    return this.getDevices().some(cachedDevice => this.compareDevices(device, cachedDevice));
+  }
+
+  /**
+   * Look through the cached output devices and match based on
+   *  the passed in output deviceId. Returns the found device
+   *  or `undefined` if the device could not be found.
+   * @param id output deviceId
+   */
+  findCachedOutputDeviceById (id?: string): MediaDeviceInfo | undefined {
+    return this.getState().outputDevices.find(d => d.deviceId === id);
+  }
+
+  /**
+   * This will remove all media listeners, stop any existing media,
+   *  and stop listening for device changes.
+   *
+   * WARNING: calling this effectively renders this SDK
+   *  instance useless. A new instance will need to be
+   *  created after this has been called.
+   */
+  destroy () {
+    this.removeAllListeners();
+    window.navigator.mediaDevices.removeEventListener('devicechange', this.handleDeviceChange.bind(this));
```

Comment:
I'm not sure this works. In fact, I think if you call `addEventListener` with a bound function, you need to keep track of it in order to remove it since `.bind()` creates a new function.

Reply:
see https://stackoverflow.com/questions/11565471/removing-event-listener-which-was-added-with-bind

Reply:
Oooh I didn't realize that. I will update it.